storage/swift_storage.py

```diff
@@ -1,253 +1,254 @@
-import functools
 import mimetypes
 import os
-import urllib
-import urlparse
+from urllib.parse import parse_qsl, urljoin, urlparse
 
-import pyrax
+from keystoneauth1 import session  # type: ignore
+from keystoneauth1.identity import v2  # type: ignore
+from keystoneauth1.exceptions.http import Forbidden, Unauthorized  # type: ignore
+import swiftclient  # type: ignore
+from swiftclient.exceptions import ClientException  # type: ignore
+from typing import Any, BinaryIO, Callable, cast, Dict, List
+from typing import Optional, Sequence, Tuple, Type, TypeVar
 
 from . import retry
-from .storage import Storage, InvalidStorageUri, timeout, register_storage_protocol, \
-    _LARGE_CHUNK, DEFAULT_SWIFT_TIMEOUT
+from .storage import _LARGE_CHUNK, DEFAULT_SWIFT_TIMEOUT, register_storage_protocol, Storage
+
+
+def register_swift_protocol(
+        scheme: str, auth_endpoint: str) -> Callable[[Type["SwiftStorage"]], Type["SwiftStorage"]]:
+    def wrapper(cls: Type["SwiftStorage"]) -> Type["SwiftStorage"]:
+        cls.auth_endpoint = auth_endpoint
+        return cast(Type[SwiftStorage], register_storage_protocol(scheme)(cls))
+    return wrapper
+
+
+class SwiftStorageError(Exception):
+
+    def __init__(self, message: str, do_not_retry: bool = False) -> None:
+        super().__init__(message)
+        self.do_not_retry = do_not_retry
+
+
+T = TypeVar("T")
+
+
+def retry_swift_operation(error_str: str, fn: Callable[..., T], *args: Any, **kwargs: Any) -> T:
+
+    def wrap_swift_operations() -> T:
+        try:
+            return fn(*args, **kwargs)
+        except Forbidden:
+            raise SwiftStorageError(
+                f"Keystone authorization returned forbidden access {error_str}",
+                do_not_retry=True)
+        except Unauthorized:
+            raise SwiftStorageError(
+                f"Keystine authorization return unauthorized access {error_str}",
+                do_not_retry=True)
+        except ClientException as swift_exception:
+            raise SwiftStorageError(
+                f"Unable to perform swift operation {error_str}: {swift_exception}",
+                do_not_retry=True)
+
+    try:
+        return retry.attempt(wrap_swift_operations)
+    except SwiftStorageError:
+        raise
+    except Exception as exc:
+        raise SwiftStorageError(f"Failure retrieving object: {exc}")
```

Comment:
I think this should be `raise SwiftStorageError(...) from exc`, which will include both stack traces in any subsequent logs, and should make failures easier to debug.